### ISSUE TITLE
[CI] ClassificationIT.testWithOnlyTrainingRowsAndTrainingPercentIsFifty_DependentVariableIsBoolean

### ISSUE THREAD

**Issue by andreidan** [elastic/elasticsearch#60759]
There are several tests from `ClassificationIT` failing (they might be correlated) - see https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+7.x+release-tests/263/ 

I've also seen quite a few failures of `ClassificationIT` over the last few months (none have been tracked though).

**Build scan**:
https://gradle-enterprise.elastic.co/s/efsbpr5sbnzwk

**Repro line**:
```./gradlew ':x-pack:plugin:ml:qa:native-multi-node-tests:integTest' --tests "org.elasticsearch.xpack.ml.integration.ClassificationIT.testWithOnlyTrainingRowsAndTrainingPercentIsFifty_DependentVariableIsBoolean" \
  -Dtests.seed=158A5379F8D00EFB \
  -Dtests.security.manager=true \
  -Dbuild.snapshot=false \
  -Dtests.jvm.argline="-Dbuild.snapshot=false" \
  -Dtests.locale=lt \
  -Dtests.timezone=America/Hermosillo \
  -Druntime.java=8
```

**Reproduces locally?**: No

**Applicable branches**: 7.x

**Failure history**:

**Failure excerpt**:
```
java.lang.AssertionError: Stats were: {"id":"classification_training_percent_is_50_boolean","state":"analyzing","progress":[{"phase":"reindexing","progress_percent":100},{"phase":"loading_data","progress_percent":100},{"phase":"feature_selection","progress_percent":100},{"phase":"coarse_parameter_search","progress_percent":100},{"phase":"fine_tuning_parameters","progress_percent":100},{"phase":"final_training","progress_percent":0},{"phase":"writing_results","progress_percent":0},{"phase":"inference","progress_percent":0}],"data_counts":{"training_docs_count":150,"test_docs_count":0,"skipped_docs_count":0},"memory_usage":{"timestamp":1596639663246,"peak_usage_bytes":17924,"status":"ok"},"node":{"id":"qTn19sC3Qju6ZOgwo7eDqw","name":"integTest-2","ephemeral_id":"oOFqlRycRB6L868ehr6p6w","transport_address":"127.0.0.1:34239","attributes":{"testattr":"test","ml.machine_memory":"101203308544","ml.max_open_jobs":"20","xpack.installed":"true","transform.node":"true"}},"assignment_explanation":""}
Expected: <stopped>
     but: was <analyzing>
	at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:18)
	at org.junit.Assert.assertThat(Assert.java:956)
	at org.elasticsearch.xpack.ml.integration.MlNativeDataFrameAnalyticsIntegTestCase.assertIsStopped(MlNativeDataFrameAnalyticsIntegTestCase.java:207)
	at org.elasticsearch.xpack.ml.integration.MlNativeDataFrameAnalyticsIntegTestCase.lambda$waitUntilAnalyticsIsStopped$0(MlNativeDataFrameAnalyticsIntegTestCase.java:157)
	at org.elasticsearch.test.ESTestCase.assertBusy(ESTestCase.java:951)
	at org.elasticsearch.xpack.ml.integration.MlNativeDataFrameAnalyticsIntegTestCase.waitUntilAnalyticsIsStopped(MlNativeDataFrameAnalyticsIntegTestCase.java:157)
	at org.elasticsearch.xpack.ml.integration.MlNativeDataFrameAnalyticsIntegTestCase.waitUntilAnalyticsIsStopped(MlNativeDataFrameAnalyticsIntegTestCase.java:153)
	at org.elasticsearch.xpack.ml.integration.ClassificationIT.testWithOnlyTrainingRowsAndTrainingPercentIsFifty(ClassificationIT.java:275)
	at org.elasticsearch.xpack.ml.integration.ClassificationIT.testWithOnlyTrainingRowsAndTrainingPercentIsFifty_DependentVariableIsBoolean(ClassificationIT.java:354)
```

**Comment by droberts195** [comment]
There were a similar set of failures in https://gradle-enterprise.elastic.co/s/eikzb6eka4hk4

The `ClassificationIT.testWithOnlyTrainingRowsAndTrainingPercentIsFifty_DependentVariableIsBoolean` failed with this:

```
java.lang.AssertionError: Stats were: {"id":"classification_training_percent_is_50_boolean","state":"starting","progress":[{"phase":"reindexing","progress_percent":1},{"phase":"loading_data","progress_percent":0},{"phase":"feature_selection","progress_percent":0},{"phase":"coarse_parameter_search","progress_percent":0},{"phase":"fine_tuning_parameters","progress_percent":0},{"phase":"final_training","progress_percent":0},{"phase":"writing_results","progress_percent":0},{"phase":"inference","progress_percent":0}],"data_counts":{"training_docs_count":0,"test_docs_count":0,"skipped_docs_count":0},"memory_usage":{"peak_usage_bytes":0,"status":"ok"},"node":{"id":"_cUVYHvfR9WfE9JsYwk_6Q","name":"integTest-2","ephemeral_id":"T2Py-4dsRyqVIYV2kHqN9Q","transport_address":"127.0.0.1:56721","attributes":{"testattr":"test","ml.machine_memory":"101104721920","ml.max_open_jobs":"20","xpack.installed":"true","transform.node":"true"}},"assignment_explanation":""}
   Expected: <stopped>
   but: was <starting> 
```

Then 6 more tests failed afterwards with a timeout on cluster health during their test cleanup.  The reason for that is that one of the nodes dies due to an assertion error:

```
[2020-08-10T12:50:27,133][ERROR][o.e.b.ElasticsearchUncaughtExceptionHandler] [integTest-0] fatal error in thread [elasticsearch[integTest-0][ml_job_comms][T#1]], exiting
java.lang.AssertionError: null
        at org.elasticsearch.xpack.core.ml.inference.trainedmodel.inference.EnsembleInferenceModel.innerInfer(EnsembleInferenceModel.java:139) ~[?:?]
        at org.elasticsearch.xpack.core.ml.inference.trainedmodel.inference.EnsembleInferenceModel.infer(EnsembleInferenceModel.java:123) ~[?:?]
        at org.elasticsearch.xpack.core.ml.inference.trainedmodel.inference.InferenceDefinition.infer(InferenceDefinition.java:80) ~[?:?]
        at org.elasticsearch.xpack.ml.inference.loadingservice.LocalModel.inferNoStats(LocalModel.java:121) ~[?:?]
        at org.elasticsearch.xpack.ml.dataframe.inference.InferenceRunner.inferTestDocs(InferenceRunner.java:116) ~[?:?]
        at org.elasticsearch.xpack.ml.dataframe.inference.InferenceRunner.run(InferenceRunner.java:88) ~[?:?]
        at org.elasticsearch.xpack.ml.dataframe.process.AnalyticsProcessManager.runInference(AnalyticsProcessManager.java:343) ~[?:?]
        at org.elasticsearch.xpack.ml.dataframe.process.AnalyticsProcessManager.processData(AnalyticsProcessManager.java:196) ~[?:?]
        at org.elasticsearch.xpack.ml.dataframe.process.AnalyticsProcessManager.lambda$runJob$1(AnalyticsProcessManager.java:146) ~[?:?]
        at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingRunnable.run(ThreadContext.java:651) ~[elasticsearch-7.10.0-SNAPSHOT.jar:7.10.0-SNAPSHOT]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) ~[?:1.8.0_241]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) ~[?:1.8.0_241]
        at java.lang.Thread.run(Thread.java:748) [?:1.8.0_241]
```

Because the tests are against a multi-node cluster the subsequent tests all succeed running against the two nodes that are left (which is good), and the only thing that fails is the check that the cluster is fully healthy.

**Comment by droberts195** [comment]
The assertion that's failing is `assert featureNames != null && featureNames.length > 0`, so it seems that `ClassificationIT.testWithOnlyTrainingRowsAndTrainingPercentIsFifty_DependentVariableIsBoolean` can cause that sometimes.

**Comment by davidkyle** [comment]
Another failure on 7.x 

https://gradle-enterprise.elastic.co/s/aamxj7fftpyim

Similar to the analysis above subsequent tests failed because the cluster cannot get to a green state with 2 nodes 


```
java.lang.AssertionError: Stats were: {"id":"classification_training_percent_is_50_integer","state":"starting","progress":[{"phase":"reindexing","progress_percent":1},{"phase":"loading_data","progress_percent":0},{"phase":"feature_selection","progress_percent":0},{"phase":"coarse_parameter_search","progress_percent":0},{"phase":"fine_tuning_parameters","progress_percent":0},{"phase":"final_training","progress_percent":0},{"phase":"writing_results","progress_percent":0},{"phase":"inference","progress_percent":0}],"data_counts":{"training_docs_count":0,"test_docs_count":0,"skipped_docs_count":0},"memory_usage":{"peak_usage_bytes":0,"status":"ok"},"node":{"id":"7JwK-kHNQ0SRII5ZIGvn0g","name":"javaRestTest-2","ephemeral_id":"8XZaFiNoQ--_HfVw_beZGA","transport_address":"127.0.0.1:52456","attributes":{"testattr":"test","ml.machine_memory":"34359738368","ml.max_open_jobs":"20","xpack.installed":"true","transform.node":"true"}},"assignment_explanation":""}
Expected: <stopped>
     but: was <starting>
```

**Comment by benwtrent** [comment]
The interesting logs:
```
[2020-09-07T06:34:11,718][INFO ][o.e.x.m.d.i.InferenceRunner] [javaRestTest-1] [classification_training_percent_is_50_integer] Started inference on test data against model [classification_training_percent_is_50_integer-1599485651478]
[2020-09-07T06:34:11,720][DEBUG][o.e.x.c.m.i.t.i.EnsembleInferenceModel] [javaRestTest-1] rewriting features {}
[2020-09-07T06:34:11,721][DEBUG][o.e.x.c.m.i.t.i.EnsembleInferenceModel] [javaRestTest-1] detected submodel feature names []
[2020-09-07T06:34:11,721][DEBUG][o.e.x.c.m.i.t.i.TreeInferenceModel] [javaRestTest-1] rewriting features {}
[2020-09-07T06:34:11,721][DEBUG][o.e.x.m.d.i.InferenceRunner] [javaRestTest-1] Loaded inference model [LocalModel{trainedModelDefinition=org.elasticsearch.xpack.core.ml.inference.trainedmodel.inference.InferenceDefinition@68a52920, modelId='classification_training_percent_is_50_integer-1599485651478', fieldNames=[text-field, keyword-field, boolean-field, outer-field.inner-field, alias-to-nested-field, numerical-field, alias-to-keyword-field], defaultFieldMap=null, statsAccumulator=org.elasticsearch.xpack.core.ml.inference.trainedmodel.InferenceStats$Accumulator@54e136, trainedModelStatsService=org.elasticsearch.xpack.ml.inference.TrainedModelStatsService@715eaf81, persistenceQuotient=100, currentInferenceCount=0, inferenceConfig=org.elasticsearch.xpack.core.ml.inference.trainedmodel.ClassificationConfig@646a9ae6, licenseLevel=PLATINUM, trainedModelCircuitBreaker=org.elasticsearch.common.breaker.ChildMemoryCircuitBreaker@28450ef0, referenceCount=1}]
[2020-09-07T06:34:11,725][DEBUG][o.e.x.c.m.i.t.i.EnsembleInferenceModel] [javaRestTest-1] Inference called with feature names []
```

**Comment by jtibshirani** [comment]
I think we just saw the same failure on 7.9: https://gradle-enterprise.elastic.co/s/kb6rkdcgxggim. Should the fix be backported to 7.9 as well?
